### PR TITLE
[Playground] [Backend] updated the method to run code to process exam…

### PR DESCRIPTION
**Pull request by vchunikhin**
@rshamunov 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: playground/backend/datasets/dataset.json [hunk 1]
```diff
@@ -1,24 +1,22 @@
...
-      "value": "hi vlad"
-    },
-    {
-      "key": 5,
-      "value": "welcome ana"
-    }
-  ]
-}
\ No newline at end of file
+[
```
Comment (rshamunov) [inline playground/backend/datasets/dataset.json:1]: Pick more valuable dataset

File: playground/backend/cmd/server/controller.go [hunk 2]
```diff
@@ -71,46 +76,73 @@ func (controller *playgroundController) RunCode(ctx context.Context, info *pb.Ru
...
 		logger.Errorf("RunCode(): unimplemented sdk: %s\n", info.Sdk)
-		return nil, errors.InvalidArgumentError("Error during preparing", "Sdk is not implemented yet: %s", info.Sdk.String())
+		return nil, errors.InvalidArgumentError(errorTitleRunCode, "Sdk is not implemented yet: %s", info.Sdk.String())
 	}
 
 	cacheExpirationTime := controller.env.ApplicationEnvs.CacheEnvs().KeyExpirationTime()
 	pipelineId := uuid.New()
 
-	lc, err := life_cycle.Setup(info.Sdk, info.Code, pipelineId, controller.env.ApplicationEnvs.WorkingDir(), controller.env.ApplicationEnvs.PipelinesFolder(), controller.env.BeamSdkEnvs.PreparedModDir())
+	var kafkaMockCluster emulators.EmulatorMockCluster
```
Comment (eantyshev) [inline playground/backend/cmd/server/controller.go:90]: Probably, we need to encapsulate this in a function at least

File: playground/backend/internal/cloud_bucket/cloud_storage.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	if cs.client == nil {
+		return nil, errors.New("storage client not set")
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	bucket := cs.client.Bucket(bucketName)
+
+	var rc storage.Reader
```
Comment (eantyshev) [inline playground/backend/internal/cloud_bucket/cloud_storage.go:59]: If it's an interface, it holds a reference to the value inside, no need to pass it explicitly below

File: playground/backend/internal/cloud_bucket/cloud_storage.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	defer func(rc *storage.Reader) {
+		err := rc.Close()
+		if err != nil {
+			logger.Error(err.Error())
+		}
+	}(&rc)
+	result := make([]*DatasetDTO, 0, len(datasets))
+	for _, dataset := range datasets {
```
Comment (eantyshev) [inline playground/backend/internal/cloud_bucket/cloud_storage.go:67]: Could you explain what is parsed below? looks cryptic

File: playground/backend/internal/emulators/kafka.go [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+		return nil, errors.New("wrong bootstrap server value")
+	}
+
+	workTicker := time.NewTicker(pauseDuration)
+	globalTicker := time.NewTicker(globalDuration)
+	for {
+		select {
+		case <-workTicker.C:
+			if _, err = net.DialTimeout(networkType, bootstrapServers, pauseDuration); err == nil {
+				workTicker.Stop()
```
Comment (eantyshev) [inline playground/backend/internal/emulators/kafka.go:68]: These objects will be destroyed upon func returned, don't need to stop

File: playground/backend/internal/emulators/kafka.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+	defer kp.producer.Close()
+	for _, dataset := range datasets {
+		topicName := dataset.Dataset.Options[constants.TopicNameKey]
+		topic := new(string)
+		*topic = topicName
+
+		ext := filepath.Ext(dataset.Dataset.DatasetPath)
+		var entries []map[string]interface{}
+		switch ext {
+		case jsonExt:
```
Comment (eantyshev) [inline playground/backend/internal/emulators/kafka.go:121]: De-serialization needs to be encapsulated in a separate func

**Review by eantyshev — COMMENTED** [review]
Looks nice! minor comments though

**Comment by eantyshev** [comment]
```
E   ImportError: cannot import name 'create_file' from 'apache_beam.testing.test_utils' (/usr/local/lib/python3.7/site-packages/apache_beam/testing/test_utils.py)
```
something wrong with Python examples

File: playground/backend/internal/cloud_bucket/cloud_storage.go (hunk 3 above)
Comment (vchunikhin) [reply playground/backend/internal/cloud_bucket/cloud_storage.go:60]: It's not an interface, it's a struct:

```
type Reader struct {
	Attrs              ReaderObjectAttrs
	seen, remain, size int64
	checkCRC           bool   // should we check the CRC?
	wantCRC            uint32 // the CRC32c value the server sent in the header
	gotCRC             uint32 // running crc

	reader io.ReadCloser
}
```

**Pull request closed without merging**